InputDeviceManagementService

J0EK3R opened the pull request:
After **many hours** of _thinking, refactoring, moving and renaming files_ - here is my **suggestion for a generic handling of inputdevices** (like gamecontrollers, MCP server,...).

Now, there is one instance of a `InputDeviceManagerService` handling all `InputDeviceServices` (all kinds of inputdevice services like GameControllerService, Gyro, whatever) and all `InputDevices` (gamecontrollers, gyro and so on).

The central `InputDeviceManagerService` still **raises** inputdevice **events** on button press, axis movement, channel changes and connection/disconnection inputdevices.

An `InputDeviceService` handles one **special kind of inputdevices** like gamecontrollers.
(In this branch there are only InputDeviceServices for gamecontrollers).

Now there is common base class `InputDevice` that replaces the former "Controller" - there was and still is a lot of renaming to do...
All inputdevices are derived from this baseclass.

One interesting part of code is the initialisation of all registered `InputDeviceServices` in `InputDeviceManagementModule` - perhaps there is a better way...
```
builder.RegisterType<InputDeviceManagerService>().As<IInputDeviceManagerService>().As<IInputDeviceEventServiceInternal>().As<IInputDeviceEventService>().SingleInstance()
                .OnActivated(e =>
                {
                    // resolve all registered input device services to trigger their registration
                    e.Context.Resolve<IEnumerable<IInputDeviceService>>();
                });
```

Please take some time and comment.

Commit:
added ControllerManagement for iOS

cleanup code:
* removed unused properties of IGameController and GamepadControllerBase
* removed generic in GameControllerServiceBase; use IGameController

Commit:
removed commented-out code and enshure PlayerIndex is set

Commit:
split "controllerdevice" to "controller device"

Commit:
split "Controllerdevice" to "Controller device"

Commit:
codebase simplified

Commit:
split "ControllerDevice" to "Controller device"

Commit:
changed variable name "gamepad" to "controller"

Commit:
static array of ValidPlayerIndexes

Commit:
Use GCController.VendorName if available as controller's displayed name

Commit:
use HasValueChanged from base class

Commit:
use of constants BUTTON_PRESSED and BUTTON_RELEASED

Commit:
use switch-expression and constants to adjust value

Commit:
renamed and moved files

GameController -> InhputDevice

Commit:
added InputDeviceManagementService

Commit:
GameControllerService musst be registered "AsSelf"

vicocz commented on BrickController2/BrickController2.iOS/PlatformServices/GameController/GameControllerService.cs:
```diff
@@ -1,285 +1,122 @@
-            Standard,
-            Extended
-        };
+        private static readonly GCControllerPlayerIndex[] ValidPlayerIndexes = Enum.GetValues(typeof(GCControllerPlayerIndex))
```
There is definitely generic option so as:
* there is no need for casting using `Cast<GCControllerPlayerIndex>`
* some enum value cen be simply excluded using LINQ's `Except`

J0EK3R replied:
here is a more elegant solution...

```C
        private static readonly GCControllerPlayerIndex[] ValidPlayerIndexes =
            Enum.GetValues<GCControllerPlayerIndex>()
                .Except([GCControllerPlayerIndex.Unset])
                .ToArray();

```

https://github.com/J0EK3R/brickcontroller2/commit/c4fb79b538a2b84cb089178cde229483d95215ec

vicocz commented on BrickController2/BrickController2/InputDeviceManagement/DI/InputDeviceManagementModule.cs:
```diff
@@ -0,0 +1,19 @@
+                .OnActivated(e =>
+                {
+                    // resolve all registered input device services to trigger their registration
+                    e.Context.Resolve<IEnumerable<IInputDeviceService>>();
```
Why is this needed? Why do we need `RegisterInputDeviceService` method? What's wrong with resolving `IEnumerable<IInputDeviceService>` as parameter of `InputDeviceManagerService` constructor?

vicocz replied:
Still I need to check this more deeply...

J0EK3R replied:
> IEnumerable<IInputDeviceService> as parameter of InputDeviceManagerService constructor?

That was my first attempt and it crashed - it's the chicken-or-egg problem ;)

The `InputDeviceService` derivations constructor need the `InputDeviceManagerService` instance as parameter.
So you can't resolve `IEnumerable<IInputDeviceService>` as parameter of `InputDeviceManagerService` constructor.
So we first need to create the InputDeviceManagerService instance and then all the registered InputDeviceService types.

vicocz reviewed (commented):
Still out of time, therefore left few comments only.

But in general I like Input Device naming. So I guess this will be nice feature to complete, even if it's kind of refactoring only.

Commit:
fixed Typo

NotifyInputDevicessChangedAction  -> NotifyInputDevicesChangedAction

J0EK3R commented on BrickController2/BrickController2/PlatformServices/InputDevice/InputDeviceChangedEventArgs.cs:
```diff
@@ -0,0 +1,29 @@
+    public IReadOnlyCollection<IInputDevice> Items { get; }
+}
+
+public enum NotifyInputDevicessChangedAction
```
Fixed typo

https://github.com/J0EK3R/brickcontroller2/commit/2ce20d5e0e35e0770861beecf99ba1f601902103

J0EK3R commented on BrickController2/BrickController2/UI/ViewModels/ControllerTesterPageViewModel.cs:
```diff
@@ -33,69 +34,69 @@ public override void OnAppearing()
                     RaisePropertyChanged(nameof(ControllerEventList));
                     break;
-                case NotifyGameControllersChangedAction.Disconnected:
+                case NotifyInputDevicessChangedAction.Disconnected:
```
fixed Typo

https://github.com/J0EK3R/brickcontroller2/commit/2ce20d5e0e35e0770861beecf99ba1f601902103

J0EK3R commented on BrickController2/BrickController2/UI/ViewModels/ControllerTesterPageViewModel.cs:
```diff
@@ -33,69 +34,69 @@ public override void OnAppearing()
                     RaisePropertyChanged(nameof(ControllerEventList));
                     break;
-                case NotifyGameControllersChangedAction.Disconnected:
+                case NotifyInputDevicessChangedAction.Disconnected:
```
fixed typo

https://github.com/J0EK3R/brickcontroller2/commit/2ce20d5e0e35e0770861beecf99ba1f601902103

J0EK3R commented on BrickController2/BrickController2/InputDeviceManagement/InputDeviceManagerService.cs:
```diff
@@ -0,0 +1,236 @@
+                inputDevice.Stop();
+
+                // notify removal
+                RaiseInputDevicesChanged(NotifyInputDevicessChangedAction.Disconnected, inputDevice);
```
fixed typo

https://github.com/J0EK3R/brickcontroller2/commit/2ce20d5e0e35e0770861beecf99ba1f601902103

J0EK3R commented on BrickController2/BrickController2/InputDeviceManagement/InputDeviceManagerService.cs:
```diff
@@ -0,0 +1,236 @@
+    /// <summary>
+    /// raise inputdevices changed event
+    /// </summary>
+    private void RaiseInputDevicesChanged(NotifyInputDevicessChangedAction action, params IInputDevice[] inputDevices)
```
fixed typo

https://github.com/J0EK3R/brickcontroller2/commit/2ce20d5e0e35e0770861beecf99ba1f601902103

Commit:
ValidPlayerIndexes - more elegant

J0EK3R commented on BrickController2/BrickController2/InputDeviceManagement/InputDeviceManagerService.cs:
```diff
@@ -0,0 +1,236 @@
+            inputDevice.Start();
+
+            // notify adding
+            RaiseInputDevicesChanged(NotifyInputDevicessChangedAction.Connected, inputDevice);
```
fixed typo

https://github.com/J0EK3R/brickcontroller2/commit/2ce20d5e0e35e0770861beecf99ba1f601902103

Commit:
revert removal of IsControllerIdSupported - there are still bindings in xaml

J0EK3R commented on BrickController2/BrickController2/UI/ViewModels/ControllerProfilePageViewModel.cs:
```diff
@@ -86,8 +86,6 @@ public List<ControllerEventViewModel> ControllerEvents
             set { _controllerEvents = value; RaisePropertyChanged(); }
         }
 
```
There are still bindings in the view

https://github.com/J0EK3R/brickcontroller2/commit/308f7d8df014ab7ae1da4f0f1d8092db04324bac

Commit:
removed property IInputDeviceEventService.IsControllerIdSupported

TODOs in
* ControllerProfilePageViewModel
* ControllerTesterPageViewModel

J0EK3R commented on BrickController2/BrickController2/InputDeviceManagement/InputDeviceManagerService.cs:
```diff
@@ -0,0 +1,236 @@
+    /// <summary>
+    /// returns true if the service supports controller ids
+    /// </summary>
+    public bool IsControllerIdSupported => true; // to be removed later, because all inputdevice services support ControllerId now
```
done

https://github.com/J0EK3R/brickcontroller2/commit/51cad4c5c1798fcd04420a8cf088e428d0202ad9

J0EK3R commented on BrickController2/BrickController2/PlatformServices/InputDeviceService/IInputDeviceEventService.cs:
```diff
@@ -0,0 +1,26 @@
+
+        /// <summary>
+        /// returns true if the service supports controller ids
+        /// </summary>
```
done

https://github.com/J0EK3R/brickcontroller2/commit/51cad4c5c1798fcd04420a8cf088e428d0202ad9

Commit:
Removed IsControllerIdSupported property as all inputdevice services support ControllerId now

Commit:
Removed IsGrouped property because IsControllerIdSupported property always is true as all inputdevice services support ControllerId now

Commit:
renamed ControllerTesterPage* to InputDeviceTesterPage*

Commit:
renamed classes ControllerTester* to InputDeviceTester*

Commit:
renamed resources ControllerTester to InputDeviceTester

Commit:
added translation resources "InputDevice" and "InputDevices"

J0EK3R commented:
I've startet renaming all the Controller-stuff (types, methods, etc.) to match the more common "InputDevice"-naming...
Does it make sense - yes, I think so.
But maybe we loose the ability to load or shere old stuff (Database?, json serialization...?)

What do you think?

vicocz commented:
> I've startet renaming all the Controller-stuff (types, methods, etc.) to match the more common "InputDevice"-naming... Does it make sense - yes, I thing so. But maybe we loose the ability to load or shere old stuff (Database?, json serialization...?)
> 
> What do you think?

No breaking changes from persistence point of view. So JSON related names should be the same.
Honestly, I will not approve such changes.
The goal is not to make the app nicest but keep it maintained and adding reasonable features.

J0EK3R commented:
> One interesting part of code is the initialisation of all registered InputDeviceServices in InputDeviceManagementModule - perhaps there is a better way...
>
> builder.RegisterType<InputDeviceManagerService>().As<IInputDeviceManagerService>().As<IInputDeviceEventServiceInternal>().As<IInputDeviceEventService>().SingleInstance()
                .OnActivated(e =>
                {
                    // resolve all registered input device services to trigger their registration
                    e.Context.Resolve<IEnumerable<IInputDeviceService>>();
                });

It seems, there is a much better solution to instantiate objects at startup: `IStartable`

https://autofac.org/apidoc/html/5F3C297F.htm

Commit:
use of Autofac's IStartable

to ensure InputDeviceService is instantiated and started as soon as the container is built in Autofac

J0EK3R commented:
> > One interesting part of code is the initialisation of all registered InputDeviceServices in InputDeviceManagementModule - perhaps there is a better way...
> > builder.RegisterType().As().As().As().SingleInstance()
> > .OnActivated(e =>
> > {
> > // resolve all registered input device services to trigger their registration
> > e.Context.Resolve<IEnumerable>();
> > });
> 
> It seems, there is a much better solution to instantiate objects at startup: `IStartable`
> 
> https://autofac.org/apidoc/html/5F3C297F.htm

Oh, yezzz - much better:

```C
builder.RegisterType<GameControllerService>().As<IInputDeviceService>().As<IStartable>().SingleInstance(); // ensure it's started as soon as the container is built in Autofac
```


https://github.com/J0EK3R/brickcontroller2/commit/06399393632a3aa1efa4472cf7d2a8ddcaa69f3f

J0EK3R commented:
> > I've startet renaming all the Controller-stuff (types, methods, etc.) to match the more common "InputDevice"-naming... Does it make sense - yes, I thing so. But maybe we loose the ability to load or shere old stuff (Database?, json serialization...?)
> > What do you think?
> 
> No breaking changes from persistence point of view. So JSON related names should be the same. Honestly, I will not approve such changes. The goal is not to make the app nicest but keep it maintained and adding reasonable features.

I have identified these types that must retain their name and the names of their properties - they are used in database and sharing context:
* `ControllerAction`
* `ControllerEvent`
* `ControllerProfile`

I've renamed files and types `ControllerTesterPage` and `ControllerTester` and keep my fingers off the rest... ;)

J0EK3R commented:
Do you think this change from `GameControllerEventType` to `InputDeviceEventType` is a problem?
All Names aren't changed.

<img width="1367" height="683" alt="grafik" src="https://github.com/user-attachments/assets/1ed5b91a-dba7-412f-b64a-2ebfac791ab6" />

No Problem:
I've tested it - here is the **shared json** created by the Win-App (default-branch)
The `EventType` is serialized as `integer`.
```
{"ct":"bc2c","p":{"Name":"Test","ControllerProfiles":[{"Name":"Grundprofil","ControllerEvents":[{"ControllerId":"Controller 1","EventType":1,"EventCode":"Y","ControllerActions":[{"DeviceId":"MK4#Device1","MaxOutputPercent":100,"AxisActiveZonePercent":100,"MaxServoAngle":90,"StepperAngle":90,"SequenceName":""}]},{"ControllerId":"McpServer","EventType":1,"EventCode":"0","ControllerActions":[{"DeviceId":"MK4#Device1","MaxOutputPercent":100,"AxisActiveZonePercent":100,"MaxServoAngle":90,"StepperAngle":90,"SequenceName":""}]},{"ControllerId":"McpServer","EventType":1,"EventCode":"1","ControllerActions":[{"DeviceId":"MK4#Device1","Channel":1,"MaxOutputPercent":100,"AxisActiveZonePercent":100,"MaxServoAngle":90,"StepperAngle":90,"SequenceName":""},{"DeviceId":"WeDo2#48:87:2D:B3:75:0E","MaxOutputPercent":100,"AxisActiveZonePercent":100,"MaxServoAngle":90,"StepperAngle":90,"SequenceName":""}]}]}]}}
```
**Loading from database** isn't a problem, too.
Here is a part of `CREATE TABLE` of the sqlite database - `EventType` is saved as `integer`:
```
CREATE TABLE "Sequence" (
"Id" integer primary key autoincrement not null ,
"Name" varchar ,
"Loop" integer ,
...
"ControllerId" varchar ,
"EventType" integer ,
"EventCode" varchar )
```

Commit:
removed IInputDeviceManagerService.GetFirstUnusedInputDeviceNumber

Commit:
set IInputDeviceManagerService private

Commit:
moved generic to InputDeviceServiceBase<TInputDevice>

Commit:
refreshed TranslationResources.Designer caused by rebase

Commit:
adopted commit #c214753 ThumbStick gamepad buttons are not recognized [ios] (#178)

https://github.com/vicocz/brickcontroller2/commit/c2147531439a6e55d1f5c18c1bfa23daa9c75032

vicocz commented:
Just FYI I plan to start looking this week.

vicocz commented on BrickController2/BrickController2/UI/Controls/Dialogs.xaml.cs:
```diff
@@ -244,45 +245,45 @@ public async Task<GameControllerEventDialogResult> ShowGameControllerEventDialog
             }
 
-            async void gameControllerEventHandler(object sender, GameControllerEventArgs args)
+            async void gameControllerEventHandler(object sender, InputDeviceEventArgs args)
```
Nit pick" I would keep the naming related to usage - `inputDeviceEventHandler`

J0EK3R replied:
renamed

https://github.com/J0EK3R/brickcontroller2/commit/060cc76aee54940ad812c574eaa7004ff19b79d3

vicocz commented on BrickController2/BrickController2/UI/ViewModels/PlayerPageViewModel.cs:
```diff
@@ -19,7 +20,7 @@ public class PlayerPageViewModel : PageViewModelBase
         private readonly IDeviceManager _deviceManager;
         private readonly IDialogService _dialogService;
-        private readonly IGameControllerService _gameControllerService;
+        private readonly IInputDeviceEventService _gameControllerService;
```
Nit pick: Please rename the member as well.

J0EK3R replied:
renamed 

https://github.com/J0EK3R/brickcontroller2/commit/fb8ec32bdaf6d07db940967c5a6921f958484571

vicocz commented on BrickController2/BrickController2/UI/ViewModels/GameControllerGroupViewModel.cs:
```diff
@@ -1,18 +1,18 @@
 public class GameControllerGroupViewModel : ObservableCollection<GameControllerEventViewModel>, IComparable<GameControllerGroupViewModel>
 {
-    public GameControllerGroupViewModel(IGameController controller) : this (controller.ControllerId, controller)
+    public GameControllerGroupViewModel(IInputDevice controller) : this (controller.InputDeviceId, controller)
```
Nit pick: this is IMHO worth to rename as well.

J0EK3R replied:
I renamed class and properties and fixed bindings in InputDeviceTesterPage.xaml

https://github.com/J0EK3R/brickcontroller2/commit/684c6849171c0a9dce69cb6d064c380e9ade2e27

vicocz commented on BrickController2/BrickController2/UI/ViewModels/GameControllerEventViewModel.cs:
```diff
@@ -7,14 +7,14 @@ public class GameControllerEventViewModel : NotifyPropertyChangedSource
         private float _value;
 
-        public GameControllerEventViewModel(GameControllerEventType eventType, string eventCode, float value)
+        public GameControllerEventViewModel(InputDeviceEventType eventType, string eventCode, float value)
```
Rename please.

J0EK3R replied:
I renamed the class

https://github.com/J0EK3R/brickcontroller2/commit/684c6849171c0a9dce69cb6d064c380e9ade2e27

vicocz commented on BrickController2/BrickController2/UI/Services/Dialog/IDialogServer.cs:
```diff
@@ -1,9 +1,9 @@
     public interface IDialogServer : IDialogService
     {
-        IGameControllerService? GameControllerService { get; set; }
+        IInputDeviceEventService? GameControllerService { get; set; }
```
How much of work is to rename this member as well?

J0EK3R replied:
renamed

https://github.com/J0EK3R/brickcontroller2/commit/f15f46099c031f52a0f5dbab97f752340c8ec6d3

vicocz commented on BrickController2/BrickController2/Resources/TranslationResources.resx:
```diff
@@ -717,4 +714,13 @@
+    <value>Inputdevice tester</value>
+  </data>
+  <data name="InputDevice" xml:space="preserve">
+    <value>Inputdevice</value>
```
This is weired, why not "Input device"?

J0EK3R replied:
yes, renamed

https://github.com/J0EK3R/brickcontroller2/commit/f15f46099c031f52a0f5dbab97f752340c8ec6d3

vicocz commented on BrickController2/BrickController2/PlatformServices/InputDevice/InputDeviceEventArgs.cs:
```diff
@@ -0,0 +1,27 @@
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+
+namespace BrickController2.PlatformServices.InputDevice
```
In general, if this is newly added file (or git rename did not catch it) please use file scoped namepsaces.
There are other files to apply it as well.

J0EK3R replied:
Yes, in general I agree.

But this file isn't new:

<img width="686" height="276" alt="grafik" src="https://github.com/user-attachments/assets/571b16e2-6c99-4bd7-a3d0-bd1ab4d010f4" />

J0EK3R replied:
I've fixed some files

https://github.com/J0EK3R/brickcontroller2/commit/f706f363d6a8a45d499793494ca191ade7b6de8d

J0EK3R replied:
Suggestion: let's change all files to use file scoped namespaces.
...would be just one quick commit! ;)

vicocz commented on BrickController2/BrickController2/InputDeviceManagement/IInputDeviceManagerService.cs:
```diff
@@ -0,0 +1,46 @@
+    /// Register inputdevice service (i.e. gamecontroller service, MCP server service)
+    /// </summary>
+    /// <param name="inputDeviceService">inputdevice service to be registered</param>
+    void RegisterInputDeviceService(IInputDeviceService inputDeviceService);
```
I will need to debug this and all the "hacks" as `IStartable` ...

vicocz commented on BrickController2/BrickController2.WinUI/PlatformServices/DI/PlatformServicesModule.cs:
```diff
@@ -23,7 +23,7 @@ public class PlatformServicesModule : Module
     {
         builder.RegisterType<InfraredService>().As<IInfraredService>().SingleInstance();
-        builder.RegisterType<GameControllerService>().AsSelf().As<IGameControllerService>().SingleInstance();
+        builder.RegisterType<GameControllerService>().As<IInputDeviceService>().As<IStartable>().SingleInstance(); // ensure it's started as soon as the container is built in Autofac
```
May be there might be some extension generic method so as `IStartable` hack is writen once, not 3 times.

J0EK3R replied:
In my opinion it's good to see the types or interfaces all the types are registered for at this place.

vicocz replied:
Will think about it later.

Commit:
renamed gameControllerEventHandler to inputDeviceEventHandler

Commit:
renamed _gameControllerService to _inputDeviceEventService

Commit:
renamed GameControllerService  to InputDeviceEventService

Commit:
renamed resources "InputDevice" to "Input device"

Commit:
renamed classes GameControllerEventViewModel to InputDeviceEventViewModel and GameControllerGroupViewModel to InputDeviceGroupViewModel

renamed properties from Controller* to InputDevice*

Commit:
use file scoped namespaces

vicocz commented:
Thank you for your contribution :)

J0EK3R commented:
You're welcome. Thank you :)
